### DISCUSSION TITLE
How to autocomplete in VSCode?

### DISCUSSION THREAD

**Discussion by AlexHex7** [rdkit/rdkit#6413] [Q&A]
### Env

> python==3.8
> rdkit==2023.3.1

### Problem
I find some modules (*.so, as below)  can not be autocompleted.  Is there any way to solve it?

> from rdkit.Chem.rdchem import *
> from rdkit.Chem.rdCIPLabeler import *
> from rdkit.Chem.rdmolfiles import *
> from rdkit.Chem.rdmolops import *

### Similar Discussion
I find a similar discussion proposed in 2021.  How about now?

**Comment by hellmrf** [comment]
It would greatly improve experience with the package.

**Reply by ptosco** [reply]
Since release 2023.09.4 stubs for VSCode can already be generated when code is built from source by running `make stubs` (on Windows the build command would be `cmake --build . --target stubs`).

Stubs for VSCode will become part of standard `conda-forge` packages starting from release 2023.09.6, which will be available within a couple of weeks.

With the 2023.09.6 `conda-forge` packages VSCode autocompletion, documentation and type hints will be automatically available without need for additional configuration.

**Comment by greglandrum** [comment]
I don't understand what the request/problem is.

If I have the RDKit installed in a conda environment, open a python file in VS Code, and make sure that VS Code knows which conda environment to use (part of the normal configuration of VS Code's python integration), auto completion works fine. Here's an example:
![image](https://github.com/rdkit/rdkit/assets/540511/17cf67cc-f64b-4e82-b517-a60d2bde2622)

(ignore the light gray part text at the end of that line, that's copilot making a suggestion.

@hellmrf do you mean something other than this?



@ptosco's reply is about stubs provide lots of nicely formatted and useful information, but I think that's a different topic.

**Reply by greglandrum** [reply]
@ptosco has gently pointed out to me that I have manually installed a copy of the stubs, or this wouldn't work.
Apologies for my confusion.

**Reply by hellmrf** [reply]
@greglandrum how could I manually install stubs? I need to replace rdkit by a local-compiled? 

## before edit
I also have a conda environment pointed in VSCode. But I still don't get any information.

![image](https://github.com/rdkit/rdkit/assets/25724069/f86e0b2d-ed10-485b-a78c-dff0bc44cbad)

For example, `rdkit.Chem.MolFromInchi` is completed, but the more common `rdkit.Chem.MolFromSmiles` don't.

![image](https://github.com/rdkit/rdkit/assets/25724069/ab7baf14-70c8-4aea-b943-fe51a6a83d07)

**Reply by greglandrum** [reply]
You would need to do your own RDKit build.
The stubs will be available from conda in the next RDKit patch release, which should be available either over the weekend or early next week.

**Reply by hellmrf** [reply]
Great to hear that. That timing works for me, so I will just wait the release. Thank you for the attention.

**Reply by AlexHex7** [reply]
That's Great! When will the latest version be uploaded in [pypi](https://pypi.org/project/rdkit/#history)?

**Reply by greglandrum** [reply]
We don't do those builds. The question is best asked at the rdkit-pypi repo: https://github.com/kuelumbus/rdkit-pypi
